hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/file/OMDirectoryCreateRequest.java

```diff
@@ -159,8 +172,24 @@ public OMClientResponse validateAndUpdateCache(OzoneManager ozoneManager,
             FILE_ALREADY_EXISTS);
       } else if (omDirectoryResult == DIRECTORY_EXISTS_IN_GIVENPATH ||
           omDirectoryResult == NONE) {
-        dirKeyInfo = createDirectoryKeyInfo(ozoneManager, omBucketInfo,
-            volumeName, bucketName, keyName, keyArgs, transactionLogIndex);
+        dirKeyInfo = createDirectoryKeyInfo(ozoneManager, keyName, keyArgs,
+            baseObjId, transactionLogIndex);
+
+        for (String missingKey : missingParents) {
+          LOG.debug("missing parent {}", missingKey);
+          // what about keyArgs for parent directories? TODO
+          OmKeyInfo parentKeyInfo = createDirectoryKeyInfoNoACL(ozoneManager,
+              missingKey, keyArgs, baseObjId + objectCount,
+              transactionLogIndex);
```

Comment:
How are we calculating the objectID here? Is this to ensure that the objectIDs are unique?
I am not sure what the effects would be as we set the transactionID as the objectID everywhere else.

Reply:
the base object ID for every create is the transaction log index left shifted by 8 bits (tx log index << 8). OMFileRequest.getObjIdFromTxId() implements this logic. I have replaced all 3 places in the code where a object id is generated, with this call. intermediate directories are created iteratively by doing a +1 starting from the base object ID. that will guarantee each object id is unique - given that the base object id starts with the unique transaction log index.

I will put in asserts to ensure there is no overflow.

Reply:
We could still end up with the same objectID for keys and buckets/ volumes.

For example, Lets say Trxn 1 creates dir1/dir2/dir3. So objectID for dir1 would be 1 << 8 + 1 i.e. 257.
Now let's say Trxn 257 comes in and creates bucket1. Then, dir1 and bucket1 would have the same objectID.

Let's change the objectID for volumes, buckets and prefixInfo's too.

Reply:
sure, will use the same object id scheme for volume and bucket.